packages/playground/playground.php

```diff
@@ -0,0 +1,215 @@
+		return;
+	}
+	wp_enqueue_style('playground', plugin_dir_url(__FILE__) . 'assets/css/playground.css', [], PLAYGROUND_VERSION);
+	wp_register_script('playground', plugin_dir_url(__FILE__) . 'assets/js/playground.js', [], PLAYGROUND_VERSION, true);
+	wp_localize_script('playground', 'playground', [
+		'zipUrl' => esc_url(get_download_page_url()),
+		'wpVersion' => PLAYGROUND_WP_VERSION,
+		'phpVersion' => PLAYGROUND_PHP_VERSION,
+		'playgroundPackageUrl' => apply_filters(
+			'playground_package_url',
+			esc_url('https://playground.wordpress.net/client/index.js'),
+		),
+		'playgroundRemoteUrl' => apply_filters(
+			'playground_remote_url',
+			esc_url('https://playground.wordpress.net/remote.html'),
+		),
+		'blueprint' => isset($_GET['blueprintUrl']) ? get_blueprint_from_url(esc_url($_GET['blueprintUrl'])) : false,
+	]);
+	wp_enqueue_script('playground');
+}
+
+/**
+ * Get the URL to download the playground package.
+ *
+ * @return string The URL to download the playground package.
+ */
+function get_download_page_url()
+{
+	return add_query_arg(
+		[
+			'download' => 1,
+		],
+		admin_url('admin.php?page=' . PLAYGROUND_ADMIN_PAGE_SLUG)
+	);
+}
+
+/**
+ * Collect the WordPress package and package it into a zip file.
+ */
+function plugins_loaded()
+{
+	if (!is_admin()) {
+		return;
+	}
+	if (!current_user_can(ADMIN_PAGE_CAPABILITY)) {
+		return;
+	}
+
+	global $pagenow;
+	if ('admin.php' !== $pagenow) {
+		return;
+	}
+
+	if (!isset($_GET['page']) || PLAYGROUND_ADMIN_PAGE_SLUG !== $_GET['page']) {
+		return;
+	}
+	if (!isset($_GET['download'])) {
+		return;
+	}
+
+	zip_collect();
+	wp_die();
+}
+
+/**
+ * Add the WordPress Playground page to the Tools menu.
+ */
+function plugin_menu()
+{
+	add_submenu_page(
+		'tools.php',
+		'WordPress Playground',
+		'Sandbox Site',
+		ADMIN_PAGE_CAPABILITY,
+		PLAYGROUND_ADMIN_PAGE_SLUG,
+		__NAMESPACE__ . '\render_playground_page',
+		NULL
+	);
+}
+
+/**
+ * Get Blueprint from URL
+ *
+ * @param string $url The URL to fetch the blueprint from.
+ * @return array|false The blueprint, or false if the request failed.
+ */
+function get_blueprint_from_url($url)
+{
+	$response = wp_safe_remote_get($url);
+	if (is_wp_error($response)) {
+		return false;
+	}
+	$blueprint = json_decode(wp_remote_retrieve_body($response), true);
+	return $blueprint;
+}
+
+/**
+ * Render the WordPress Playground page.
+ */
+function render_playground_page()
+{
+	if (isset($_GET['download'])) {
+		return;
+	}
+	include __DIR__ . '/templates/playground-page.php';
+}
+
+/**
+ * Add a "Preview Now" button to the plugin install screen.
+ *
+ * @param array $action_links An array of plugin action links.
+ * @param array $plugin The plugin data.
+ * @return array The modified array of plugin action links.
+ */
+function plugin_install_action_links($action_links, $plugin)
+{
+	if (empty($plugin['blueprints'])) {
+		return $action_links;
+	}
+
+	$blueprint = $plugin['blueprints'][0];
```

Comment:
Oooh this is quite tricky!

We can display the Preview button on every plugin, whether it has a Blueprint or not. When you install a plugin, you don't get any initial setup other than what the plugin does on activation. It makes sense to reproduce that experience here.

Then, what should we do when the plugin ships a Blueprint? My gut says _ignore the Blueprint_.

Why?

These Blueprints are written with a vanilla WordPress site in mind. They'll typically create dummy content, change site options, and so on to reach a minimal preview environment. In some cases, like bbpress, that makes a lot of sense even in the context of your site as it doesn't have any forums in place. In other cases, that would only add confusing bloat like lorem ipsum pages when your site already has pages.

Then, applying a Blueprint means the preview experience will be very different from the Installing experience. At best, you won't get the setup wizard. At worst, you'll think "what the... is this plugin going to create 50 dummy posts when I install it? I better go for another one".

So I suggest let's ignore these Blueprints and open a new issue to discuss that with @tellyworth and @StevenDufresne.

Reply:
I removed the need for blueprints in [3bdd04a](https://github.com/WordPress/playground-tools/pull/154/commits/3bdd04a1c6584cd4ef442b9a4ce88e8be0c25b87)